src/core/ext/filters/client_channel/resolver/xds/xds_resolver.cc

```diff
@@ -294,34 +338,137 @@ XdsResolver::XdsConfigSelector::XdsConfigSelector(
+  if (it != vhost.typed_per_filter_config.end()) return &it->second;
+  // Check Route.
+  it = route.typed_per_filter_config.find(instance_name);
+  if (it != route.typed_per_filter_config.end()) return &it->second;
+  // Check ClusterWeight, if any.
+  if (cluster_weight != nullptr) {
+    it = cluster_weight->typed_per_filter_config.find(instance_name);
+    if (it != cluster_weight->typed_per_filter_config.end()) return &it->second;
+  }
+  // Not found.
+  return nullptr;
+}
+
 grpc_error* XdsResolver::XdsConfigSelector::CreateMethodConfig(
-    RefCountedPtr<ServiceConfig>* method_config, const XdsApi::Route& route) {
-  grpc_error* error = GRPC_ERROR_NONE;
+    const XdsApi::Route& route,
+    const XdsApi::Route::ClusterWeight* cluster_weight,
+    RefCountedPtr<ServiceConfig>* method_config) {
   std::vector<std::string> fields;
+  // Set timeout.
   if (route.max_stream_duration.has_value() &&
       (route.max_stream_duration->seconds != 0 ||
        route.max_stream_duration->nanos != 0)) {
     fields.emplace_back(absl::StrFormat("    \"timeout\": \"%d.%09ds\"",
                                         route.max_stream_duration->seconds,
                                         route.max_stream_duration->nanos));
   }
+  // Handle xDS HTTP filters.
+  grpc_channel_args* args = grpc_channel_args_copy(resolver_->args_);
+  for (const auto& http_filter : resolver_->current_listener_.http_filters) {
+    // Stop at the router filter.  It's a no-op for us, and we ignore
+    // anything that may come after it, for compatibility with Envoy.
+    if (http_filter.config.config_proto_type_name ==
+        kXdsHttpRouterFilterConfigName) {
+      break;
+    }
+    // Find filter.  This is guaranteed to succeed, because it's checked
+    // at config validation time in the XdsApi code.
+    const XdsHttpFilterImpl* filter_impl =
+        XdsHttpFilterRegistry::GetFilterForType(
+            http_filter.config.config_proto_type_name);
+    GPR_ASSERT(filter_impl != nullptr);
+    // Allow filter to add channel args that may affect service config
+    // parsing.
+    args = filter_impl->ModifyChannelArgs(args);
+    // Find config override, if any.
+    const XdsHttpFilterImpl::FilterConfig* config_override =
+        FindFilterConfigOverride(http_filter.name,
+                                 resolver_->current_virtual_host_, route,
+                                 cluster_weight);
+    // Generate service config for filter.
+    auto method_config_field =
+        filter_impl->GenerateServiceConfig(http_filter.config, config_override);
+    if (!method_config_field.ok()) {
+      return GRPC_ERROR_CREATE_FROM_COPIED_STRING(
+          absl::StrCat("failed to generate method config for HTTP filter ",
+                       http_filter.name, ": ",
+                       method_config_field.status().ToString())
+              .c_str());
+    }
+    fields.emplace_back(*method_config_field);
```

Comment:
I found a strange edge case here, some http filters might be created more than once: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/decompressor_filter#using-different-decompressors-for-requests-and-responses

Should we come up with a way to let service config work? It's weird, if we allow custom filters, users might set several fault injection filters in a row (this is supported by Envoy). Should we allow it?

Reply:
After thinking this through for a while, I've come up with an approach that I think will be both minimally invasive and will avoid per-call overhead.

I've added a function to the channel stack API called `grpc_channel_stack_filter_instance_number()`, which a given filter can use in its `init_channel_elem()` method to determine how many other instances of the same filter exist earlier in the same stack.  For any xDS HTTP filter, the service config parser can contain a vector, where each element in the vector should be used by each successive instance of that filter in the stack.  The individual filter instances can then use their position in the stack to determine the offset in the config vector for its config.

Please let me know if you have any questions about this.  Thanks!

Reply:
I've just tweaked the API for returning the service config contents to provide a way for the configs for multiple instances of the same filter to be combined.